Code/GraphMol/Atom.cpp

```diff
@@ -621,6 +621,39 @@ double Atom::getMass() const {
   }
 }
 
+bool Atom::hasValenceViolation() const {
+  auto any_of = [](auto &vec, const auto &pred) {
+    return std::any_of(vec.begin(), vec.end(), pred);
+  };
+
+  // Ignore dummy atoms, query atoms, or atoms attached to query bonds
+  auto atomic_number = getAtomicNum();
+  auto bonds = getOwningMol().atomBonds(this);
+  auto is_query = [](auto b) { return b->hasQuery(); };
+  if (atomic_number == 0 || hasQuery() || any_of(bonds, is_query)) {
+    return false;
+  }
+
+  // Account for charge via the isolobal analogy by modifying the atomic
+  // number used to look up allowed valences
+  auto valence_lookup = atomic_number - getFormalCharge();
+  // Because the look up accounted for charge, we can ignore that contribution
+  // to the current valence; however, still account for unpaired electrons
+  int current_valence = getTotalValence() + getNumRadicalElectrons();
+  // If no such element to look up, reset to the original atomic number
+  // and modify the current valence to account for charge
+  if (valence_lookup < 1 || valence_lookup > 118) {
+    valence_lookup = atomic_number;
+    current_valence -= getFormalCharge();
+  }
+
+  // -1 indicates that any valence is permitted
+  const auto *table = RDKit::PeriodicTable::getTable();
+  auto allowed_valence_list = table->getValenceList(valence_lookup);
+  auto is_allowed = [&](auto x) { return x == -1 || x == current_valence; };
```

Comment:
Since we know that if -1 is present it will always be the last entry in the valence list, and that the valence list is sorted, couldn't this just be:
```
return (allowed_valence_list.back() !=-1 && currentValence>allowed_valence_list.back())
```

Reply:
Doesn't checking current_valence against the last value not work? For something like P, don't I want to check that current_valence is explicitly one of 3 or 5 or 7 (where 0-2 or 4 or 6 are invalid)
```
15	P	0.75	0.89	1.95	30.974	5	31	30.97376163	3	5	7 \n \
```

I could also do this if we think it's clearer:
```
if (allowed_valence_list.back() == -1) {
    return false;
}
return std::find(allowed_valence_list.begin(), allowed_valence_list.end(),
                  current_valence) == allowed_valence_list.end();
```